### ISSUE TITLE
Centralize GE portable downloads

### ISSUE THREAD

**Issue by mast-eu** [gitextensions/gitextensions.extensibility#1]
I propose to use this repo for all scripts and resources that are common to all plugins.
The current approach simply doesn't scale well with many plugin repos.
```
C:\projects
    \gitxtensions.pluginA
        \build, \test, \src etc (the truely plugin-specific directories)
        \references (GitExtensions portable used for linking and debugging)
        \tools
            \Download-GitExtensions.ps1 (used to create ..\references)

    \gitxtensions.pluginB
        \build, \test, \src etc (the truely plugin-specific directories)
        \references (GitExtensions portable used for linking and debugging)
        \tools
            \Download-GitExtensions.ps1 (used to create ..\references)
```
We'll have n identical copies of \tools\Download-GitExtensions.ps1 and n identical downloads of GE portable in every plugin repo.
I want to move https://github.com/gitextensions/gitextensions.pluginmanager/blob/master/tools/Download-GitExtensions.ps1 here and centralize the GE portable here, too. The directory structure would be:
```
C:\projects
    \gitxtensions.extensibility
        \CommonReferences (single GitExtensions portable used for linking and debugging)
        \CommonTools
            \Download-GitExtensions.ps1 (single script to create ..\CommonReferences)

    \gitxtensions.pluginA
        \build, \test, \src etc (the truely plugin-specific directories)

    \gitxtensions.pluginB
        \build, \test, \src etc (the truely plugin-specific directories)
```
**The advantages:**
- avoid maintaining copies of PowerShell scripts for the same task in many repos
- reduce download bandwidth and disk space, since only 1 copy of GE portable is needed
- can test multiple plugins in the same instance of GE (although probably only one with the debugger attached)

**The downsides:**
- for working on any plugin, you will need to clone 2 repositories (`gitextensions.extensibility` and `gitextensions.pluginXYZ`)
- the repos must reside within a common root directory (e.g. `C:\projects`)

**Comment by mast-eu** [comment]
@gitextensions/git-extensions-plugins : Any thoughts on this?

**Comment by maraf** [comment]
I would like to have the required scripts as a nuget package that plugin references. My idea is that when you create a new plugin project, you must add reference to `GitExtensions.Plugins` and this reference ensures, in some way, that you can hit F5 and GE will load with your new plugin.

> * reduce download bandwidth and disk space, since only 1 copy of GE portable is needed

Whether centralize downloads or not, I'm not sure. Right now, my plugins have references to this folder, because sometimes I'm "hacking" things, that are not exposed / meant to be extended. This may change in the future, but until that time, we may need this dirty way.

**Comment by RussKie** [comment]
Can submodules help?

**Comment by mast-eu** [comment]
Both, submodule and nuget package, would resolve my first concern (having only one common source for shared Powershell scripts).

Both would fail to resolve my second concern (avoid a new download of GE portable for each plugin). But this is minor, let's put it aside for simplicity.

@maraf : are you familiar with a way to achieve this with nuget packages? By default packages on my PC are downloaded to `UserData\.nuget\<PackageName>`. I'm not sure if this is always the same though. How can we reliably retrieve this path from within the VS solution, in order to use it in the pre-build event?
In case you already have a solution, let's use it. Otherwise I can set it up with a submodule and we'll start from there.

**Comment by maraf** [comment]
> By default packages on my PC are downloaded to `UserData\.nuget\<PackageName>`. I'm not sure if this is always the same though.

I think I have found somewhere that configuration for this exists, but the time I tried it, it didn't worked (so I create a symlink on my machine).

> In case you already have a solution, let's use it.

I don't have any, but am curious to find out.

> Both would fail to resolve my second concern (avoid a new download of GE portable for each plugin). But this is minor, let's put it aside for simplicity.

We can use MSBuild property to setup a common location where to look for GE instance. There is no need to have it under current plugin root folder.

**Comment by mast-eu** [comment]
Ok, I will prepare a PR based on submodules to resolve the immediate issue of having scripts in a single location.

Obviously, since we will need a nuget package anyhow to distribute the APIs DLLs, the scripts might still get integrated there later on.

**Comment by RussKie** [comment]
> We can use MSBuild property to setup a common location where to look for GE instance. There is no need to have it under current plugin root folder.

For the an installed version the path is in the registry
https://github.com/gitextensions/gitextensions/blob/4c7c06401871366fb03af3ad8ba45ca54f6bc547/GitCommands/Settings/AppSettings.cs#L164-L178

**Comment by maraf** [comment]
I have made a progress with GE download script inside nuget. It's commited in [BundleBackuper](https://github.com/maraf/GitExtensions.BundleBackuper/commit/01c2877a12ec746038b3b35caba6d7c4b8151390).

Key points:
- Plugins csproj only needs to include targets from Extensibility package (which is already done by nuget).
- By default, portable GE will be downloaded and extracted somewhere inside current repository. But user/developer can set msbuild property (temporal name) `GitExtensionsExtractPath` to use a single folder for all his plugins.
- Also (before stable API) he can set version and provider (`GitExtensionsReferenceVersion` and `GitExtensionsReferenceSource`).
- For adding a reference to a GE assembly (those not inside package, but required to hack something outside of API) he uses `<HintPath>$(GitExtensionsPath)\ResourceManager.dll</HintPath>`.
- For debugging, he declares `launchSettings.json` with `"executablePath": "$(GitExtensionsPath)\\GitExtensions.exe"`.

As my key points are quite long, one more wrap up:

**Easiest way:**
- Add package `GitExtensions.Extensibility`, update `launchSettings.json` and everything is set-up.

**Customizations**
- Override GE related msbuild properties.
- Add custom references pointing to downloaded portable GE.

<hr>

Tomorrow I'm going to try to wrap msbuild targets into a nuget package.

**Comment by mast-eu** [comment]
> I have made a progress with GE download script inside nuget. It's commited in [BundleBackuper](https://github.com/maraf/GitExtensions.BundleBackuper/commit/01c2877a12ec746038b3b35caba6d7c4b8151390).

> Tomorrow I'm going to try to wrap msbuild targets into a nuget package.

Great. Could you kindly prepare a PR when you're done?

**Comment by maraf** [comment]
> Great. Could you kindly prepare a PR when you're done?

Sure, I just wanted to share the progress.

**Comment by mast-eu** [comment]
@maraf Do you have news regarding this issue?

Sorry, I do not want to stress. I just want to know whether I should try to free some time for completing PR #7 or whether you are planning to provide a PR with similar functionality.

**Comment by maraf** [comment]
@mast-eu I'm a bit overwhelmed by other projects, but I might be able to complete tests with nuget package by the end of the week.

**Comment by mast-eu** [comment]
No need to hurry, just wanted to be sure that we're not working contemporary in potentially different directions.

I'm pretty busy myself right now, because 2 weeks ago was the initial release of my long-term project "child 2.0" :baby: :baby_bottle: 

Maybe we should start to systematically assign issues to who's actually working on it, what do you think? This way it is clear who's in charge of what, even if it might take them some time.

**Comment by RussKie** [comment]
I was hesitant to ask about your personal project. Congratulations!!!

**Comment by maraf** [comment]
@mast-eu Congrats!! It's a big upgrade from the previous version 😄 

> Maybe we should start to systematically assign issues to who's actually working on it, what do you think? This way it is clear who's in charge of what, even if it might take them some time.

I definitely agree. Also I'm not sure if we have a roadmap for this repository. I know you created one for moving plugins out of the main repo.

**Comment by mast-eu** [comment]
> Congratulations!!!

> Congrats!! 

Thank you guys :smile: 

> I'm not sure if we have a roadmap for this repository.

I think a dedicated roadmap for this repo is a maybe bit of overkill, at least for the moment (or to say it with other words: I wouldn't know what to write in there ... :rofl: )
However, you're right that issues should be prioritized. Therefore, I added milestones to this repo. 
- [1.0.0](https://github.com/gitextensions/gitextensions.extensibility/milestone/1) for issues that need to be implemented / agreed on before the first official release.
- [1.1.0](https://github.com/gitextensions/gitextensions.extensibility/milestone/2) for issues that are not strictly mandatory, but still very useful and should be completed in the near future.
- [Backlog](https://github.com/gitextensions/gitextensions.extensibility/milestone/3) for everything else, that might be completed at a later stage.

Do you think this suits the need?

**Comment by RussKie** [comment]
> Also I'm not sure if we have a roadmap for this repository. I

In my mind this repo was meant to become the "glue" project between the GE and a plugin, e.g. a replacement for `GitUIPluginInterfaces` assembly.
The assembly was to get signed and published to a Nuget feed (nuget.org or GH - doesn't matter; can go to both, if necessary).
Then consumed by GE and any plugins.

**Comment by maraf** [comment]
@mast-eu Milestones are enough. Thanks

**Comment by mast-eu** [comment]
Closed in #11
